File: lib/fc/type_util.rb
# -*- coding: utf-8 -*-

module Fc

  ############################################
  # 型関係のユーティリティ
  ############################################
  module TypeUtil

    module_function

    def guess_type( type, val )
      if type
        compatible_type( type, val.type )
      else
        val.type
      end
    end

    def compatible_type?( a, b )
      return a if  a == b
      if a.kind == :int and b.kind == :int
        if a.size == b.size
          return (a.signed)?a:b
        else
          return (a.size>b.size)?a:b
        end
      elsif a.kind == :pointer and b.kind == :array and a.base == b.base
        return a
      elsif a.kind == :array and b.kind == :array and a.base == b.base and a.length == nil
        # 配列の長さを省略した場合
        return b
      elsif a.kind == :array and b.kind == :array and a.base == b.base and a.length != b.length
        return Type[[:pointer, a.base]]
      else
        return nil
      end
    end

    def compatible_type( a, b )
      r = compatible_type?( a, b )
      raise CompileError.new("not compatible type '#{a}' and '#{b}'") unless r
      r
    end

  end
end

File: internal/fc/type_util.go
package fc

// lib/fc/type_util.rb の移植。

import "fmt"

// GuessType は TypeUtil.guess_type 相当。
func GuessType(typ *Type, val Operand) *Type {
	if typ != nil {
		return CompatibleType(typ, ValType(val))
	}
	return ValType(val)
}

// CompatibleTypeOk は TypeUtil.compatible_type? 相当 (互換性がなければ nil)。
func CompatibleTypeOk(a, b *Type) *Type {
	if a == b {
		return a
	}
	if a.Kind == "int" && b.Kind == "int" {
		if a.Size == b.Size {
			if a.Signed {
				return a
			}
			return b
		}
		if a.Size > b.Size {
			return a
		}
		return b
	} else if a.Kind == "pointer" && b.Kind == "array" && a.Base == b.Base {
		return a
	} else if a.Kind == "array" && b.Kind == "array" && a.Base == b.Base && a.Length < 0 {
		// 配列の長さを省略した場合
		return b
	} else if a.Kind == "array" && b.Kind == "array" && a.Base == b.Base && a.Length != b.Length {
		return TypeOf([]any{Sym("pointer"), a.Base})
	}
	return nil
}

// CompatibleType は TypeUtil.compatible_type 相当 (互換性がなければ CompileError)。
func CompatibleType(a, b *Type) *Type {
	r := CompatibleTypeOk(a, b)
	if r == nil {
		panic(&CompileError{Msg: fmt.Sprintf("not compatible type '%s' and '%s'", a, b)})
	}
	return r
}
